src/ir/module-utils.cpp

```diff
@@ -198,10 +198,92 @@ void setIndices(IndexedHeapTypes& indexedTypes) {
   }
 }
 
+InsertOrderedSet<HeapType> getPublicTypeSet(Module& wasm) {
+  InsertOrderedSet<HeapType> publicTypes;
+
+  auto notePublic = [&](HeapType type) {
+    // TODO: Gracefully handle public types with arbitrary public children here
+    // and leave it to separate code to handle erroring or warning if
+    // optimizations would be inhibited.
+    if (type.isBasic()) {
+      return;
+    }
+    // All the rec group members are public as well.
+    for (auto member : type.getRecGroup()) {
+      if (!publicTypes.insert(type)) {
+        // We've already inserted this rec group.
+        break;
+      }
+      // Validate that this type does not have nontrivial children. We don't
+      // traverse the types to collect such children and we want to report them
+      // as errors because they might significantly inhibit optimizations if we
+      // did collect them.
+      for (auto child : member.getReferencedHeapTypes()) {
+        if (!child.isBasic()) {
+          std::string typeName = child.toString();
+          if (auto it = wasm.typeNames.find(child);
+              it != wasm.typeNames.end()) {
+            typeName = it->second.name.toString();
+          }
+          Fatal() << "nontrivial public type found during type optimizations: "
+                  << typeName;
```

Comment:
Shouldn't this only error in closed-world?

Reply:
In open world we currently do not optimize any types anyway, so for now this code path should never be reached.

Reply:
I see, makes sense.